### PR TITLE
fix: model table interactions

### LINKED ISSUE #2404: Destroying model's actions are enabled in the model's list page
I am able to interact with the deleting model on the model's list page.

1. Go to /models
2. Click on try to delete a model
3. The row for deleting model should be enabled when the destruction is in progress 

**Expected behavior**
All actions and the row for such a model should be disabled

**Screenshots**

Image

### PR DESCRIPTION
**Pull request by Ninfa-Jeon**
## Done

- Added disabled and loading states to `DataTable` rows
- Added disabled state to `ModelActions` for when an upgrade is in progress
- Moved the spinner rendered during loading (upgrading) or disabling (destroying) to be shown in the checkbox column (as per the latest [screens](https://www.figma.com/design/Rs1v3IS0oY2RK4CU4yo4Ly/25.10-JAAS-model-management---Juju?node-id=452-44876&p=f&m=dev))
- Updated styles and tests accordingly
- Added tests to improve coverage

## QA

- Pull code
- Run `dotrun clean && dotrun serve`
- Open http://0.0.0.0:8036/
- View the model table in selectable and non-selectable states
- Try deleting and upgrading models in each
- A spinner should appear under checkbox column in selectable state
- A spinner should appear in the hidden icon column in non-selectable state
- While destroying, the entire row should stay disabled
- While upgrading, the the model actions should stay disabled while the row is open to other interactions
- Try interacting with the select-all toggle when some rows are in destroying and upgrading states. It should ignore such rows.

## Details

Fixes https://github.com/canonical/juju-dashboard/issues/2404
Fixes https://warthogs.atlassian.net/browse/JUJU-10223

## Screenshots

BEFORE:
Screenshot 2026-08-12 at 5 58 34 PM
Screenshot 2026-08-12 at 5 58 55 PM

AFTER:
Screenshot 2026-08-12 at 5 56 03 PM
Screenshot 2026-08-12 at 5 56 54 PM
Screenshot 2026-08-12 at 5 56 58 PM

### CODE REVIEW

File: src/components/ModelTable/_model-table.scss [hunk 1]
```diff
@@ -13,19 +13,21 @@
...
 
-    &__name {
-      flex-shrink: 1;
-      overflow: hidden;
-    }
+  tr.is-disabled {
+    opacity: 0.5;
+    pointer-events: none;
+    user-select: none;
   }
```
Comment (andogq) [inline src/components/ModelTable/_model-table.scss]: This class is set in `useRows.tsx` (within the `DataTable` component), but it's only included in this SCSS file. If this is the approach to be used, this class must be defined in a SCSS file specific for `DataTable`.

Additionally, this is not an accessible way to 'disable' a row (pointer-events and user-select). It would be better to directly disable the interactable elements within the row (buttons/inputs/etc).

File: src/components/ModelTable/_model-table.scss [hunk 2]
```diff
@@ -13,19 +13,21 @@
...
+    width: 2rem;
+  }
 
-    &__status {
-      flex-shrink: 0;
-      margin-right: $sph--small;
-    }
+  .icon-column-header {
+    padding-top: 0.9rem;
+  }
```
Comment (andogq) [inline src/components/ModelTable/_model-table.scss]: These classes are use within `DataTable`, so should be defined there.

**Review by andogq — COMMENTED** [review]
My main question here is what does it mean for a row to be 'disabled'?

If it's intended to disable certain interactable items within the row (eg. the context menu, or a button), then the disable attribute should be directly added to them within the `renderCell` callback in the column. If it helps, the entire row data is passed as the second parameter to `renderCell`, so it can be directly referenced in every render call.

If it's intended to disable the checkbox to select a row, then that comes with its own set of problems and questions:
- If the row was selected and disabled, and the user toggles "Select all" (to clear the selection), does this checkbox change state?
- If the only row is selected and disabled, now what does the "Select all" button do?
- Is it clear to the user that a disabled row is correlated with the model being deleted (as opposed to a permission problem, a connectivity problem, etc)?

Instead, I propose the following changes:
- If a specific thing in a cell should be disabled, disable it directly in the `renderCell` method. This includes making text grey, disabling links, buttons, etc.
- Don't alter any selection logic or add any row disabling to `DataTable`. Instead, check for disabled items before doing an action with them. Eg. if we had bulk model deletion, and a model in the process of being deleted is selected, disable the bulk-delete button with a message ('cannot delete model that is already being deleted').
- Depending on Lukas' thoughts on spinner placement, it would be good to have it as an icon inline with a cell, such as below:

```jsx
          loading: column({
              header: "",
              className: "p-table__cell--icon-placeholder",
              collapse: true,
              // TODO: Determine loading state here.
              map: () => false,
              renderCell: (loading) => {
                  return loading ? <Spinner /> : null;
              }
          }),
```

image

File: src/components/ModelTable/_model-table.scss (hunk 1 above)
Comment (Ninfa-Jeon) [reply src/components/ModelTable/_model-table.scss]: done

File: src/components/ModelTable/_model-table.scss (hunk 2 above)
Comment (Ninfa-Jeon) [reply src/components/ModelTable/_model-table.scss]: done

**Comment by Ninfa-Jeon** [comment]
> If it's intended to disable certain interactable items within the row (eg. the context menu, or a button), then the disable attribute should be directly added to them within the `renderCell` callback in the column. If it helps, the entire row data is passed as the second parameter to `renderCell`, so it can be directly referenced in every render call.

The entire row needs to be rendered with 0.5 opacity besides the interactable items disabled.

> 
> If it's intended to disable the checkbox to select a row, then that comes with its own set of problems and questions:
> 
> * If the row was selected and disabled, and the user toggles "Select all" (to clear the selection), does this checkbox change state?

the disabled row is automatically deselected, and "select all" correctly ignores disabled rows

> * If the only row is selected and disabled, now what does the "Select all" button do?

The select-all checkbox becomes a no-op. I could potentially disable the "Select all" button for such a case.

> * Is it clear to the user that a disabled row is correlated with the model being deleted (as opposed to a permission problem, a connectivity problem, etc)?

Yes, because we show the message "Destroying model..." under it

> 
> Instead, I propose the following changes:
> 
> * If a specific thing in a cell should be disabled, disable it directly in the `renderCell` method. This includes making text grey, disabling links, buttons, etc.

Since the entire row needs to change its behaviour, disabling the row element from within the `DataTable` is most suitable. As opposed to per cell disabling where we wouldn't be able to set the `aria-disabled` from `renderCell`.

> * Don't alter any selection logic or add any row disabling to `DataTable`. Instead, check for disabled items before doing an action with them. Eg. if we had bulk model deletion, and a model in the process of being deleted is selected, disable the bulk-delete button with a message ('cannot delete model that is already being deleted').

As confirmed previously, the selection logic needed updates due to the latest designs.

File: src/components/ModelTable/_model-table.scss (hunk 1 above)
Comment (Ninfa-Jeon) [reply src/components/ModelTable/_model-table.scss]: splitting the disabling rules between row opacity and pointer-events/user-select has introduced somewhat redundant code. Let me know how you would like to proceed on that.

File: src/components/DataTable/_data-table.scss [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+  .is-row-disabled {
+    opacity: 0.5;
+  }
+
+  // Apply to interactive elements within a disabled row via renderCell.
+  .is-disabled {
+    pointer-events: none;
+    user-select: none;
+  }
```
Comment (andogq) [inline src/components/DataTable/_data-table.scss]: We should use the `inert` HTML property instead, as it has significantly better semantics for assistive technologies.
https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/inert

File: src/components/DataTable/_data-table.scss [hunk 4]
```diff
@@ -0,0 +1,22 @@
+.data-table {
+  .icon-column {
```
Comment (andogq) [inline src/components/DataTable/_data-table.scss:2]: Should rely on vanilla styles for icon styling in tables: https://vanillaframework.io/docs/base/tables#icons

File: src/components/DataTable/DataTable.tsx [hunk 5]
```diff
@@ -87,19 +100,19 @@ export default function DataTable<
   );
 
   return (
-    <Table>
+    <Table className="data-table">
       <Table.Header>
         {prefixHeader}
-        {selectable ? (
-          <Table.Header.Cell collapse>
+        <Table.Header.Cell className="icon-column icon-column-header">
```
Comment (andogq) [inline src/components/DataTable/DataTable.tsx:106]: Should also use the material icon styles. Also having an empty header cell is really not ideal for accessibility.

File: src/components/DataTable/useRows.tsx [hunk 6]
```diff
@@ -95,7 +105,8 @@ export default function useRows<TRow, TKey extends React.Key, TValues>({
   const fullRows = useMemo(
     () =>
       sortedRows.map((row, i) => {
-        const checked = select.selected.includes(row.key);
+        const { disabled, loading, key } = row;
```
Comment (andogq) [inline src/components/DataTable/useRows.tsx]: Calls to `isRowDisabled` and `isRowLoading` should be done here instead, to prevent re-calculating and rendering all of the rows and sorting whenever something is enabled or disabled.

File: src/components/DataTable/useRows.tsx [hunk 7]
```diff
@@ -123,22 +134,27 @@ export default function useRows<TRow, TKey extends React.Key, TValues>({
...
 
-        const label = checked ? `Select ${row.key}` : `Deselect ${row.key}`;
+        const label = checked ? `Select ${key}` : `Deselect ${key}`;
 
         return (
-          <Table.Row key={row.key}>
+          <Table.Row
+            key={key}
+            aria-disabled={disabled || undefined}
+            className={classNames({ "is-row-disabled": disabled })}
```
Comment (andogq) [inline src/components/DataTable/useRows.tsx]: This should be the `inert` attribute instead.

File: src/components/DataTable/useRows.tsx [hunk 8]
```diff
@@ -123,22 +134,27 @@ export default function useRows<TRow, TKey extends React.Key, TValues>({
...
+          >
             {prefixColumn}
-            {selectable ? (
-              <Table.Cell>
+            <Table.Cell className="icon-column">
```
Comment (andogq) [inline src/components/DataTable/useRows.tsx:146]: This should be the vanilla icon class.

File: src/components/InlineCheckbox/InlineCheckbox.tsx [hunk 9]
```diff
@@ -33,7 +33,7 @@ export default function InlineCheckbox<L extends string>({
   }, [indeterminate]);
 
   return (
-    <label className="p-checkbox">
+    <label className="p-checkbox u-no-padding--top">
```
Comment (andogq) [inline src/components/InlineCheckbox/InlineCheckbox.tsx:36]: What's the need for this to be overridden, only to be re-added manually with the icon column class?

File: src/components/ModelTable/ModelTable.tsx [hunk 10]
```diff
@@ -93,44 +99,32 @@ export default function ModelTable({ models, groupBy }: Props): JSX.Element {
...
-                  </div>
-                </div>
+                <TruncatedTooltip
+                  message={name}
+                  className={classNames({
+                    "is-disabled": isDying(uuid) || !!getUpgrade(uuid),
+                  })}
+                >
+                  <ModelDetailsLink modelName={name} qualifier={qualifier}>
+                    {model.model.name}
```
Comment (andogq) [inline src/components/ModelTable/ModelTable.tsx]: Why not use `name` here?

File: src/components/ModelTable/ModelTable.tsx [hunk 11]
```diff
@@ -93,44 +99,32 @@ export default function ModelTable({ models, groupBy }: Props): JSX.Element {
...
-                          className="models__version"
-                          modelName={name}
-                          qualifier={qualifier}
-                        />
-                      )}
-                    </TruncatedTooltip>
-                  </div>
-                </div>
+                <TruncatedTooltip
+                  message={name}
```
Comment (andogq) [inline src/components/ModelTable/ModelTable.tsx]: Would be good to include the version in the tooltip. Eg: `model name (3.6)`

File: src/components/ModelTable/ModelTable.tsx
Comment (andogq) [inline src/components/ModelTable/ModelTable.tsx:1]: All of the class names setting `is-disabled` in here are not needed. Rely on the `inert` attribute being set by the table row in the data table, and the styling associated with that.

File: src/components/ModelActions/ModelActions.test.tsx [hunk 12]
```diff
@@ -120,6 +120,21 @@ describe("ModelActions", () => {
...
+  it("displays the actions menu in disabled state", () => {
+    renderComponent(
+      <ModelActions
+        qualifier="eggman@external"
+        modelUUID="abc123"
+        modelName="test-model"
+        disabled
+      />,
+    );
+    expect(screen.getByRole("button", { name: Label.TOGGLE })).toHaveAttribute(
```
Comment (andogq) [inline src/components/ModelActions/ModelActions.test.tsx:132]: Should be testing if it's disabled: https://vitest.dev/api/browser/assertions.html#tobedisabled

File: src/components/DataTable/DataTable.test.tsx [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    it("does not render the select-all checkbox when selectable=false", () => {
+      renderComponent(<DataTable {...initialProps} selectable={false} />);
+      expect(
+        screen.queryByRole("checkbox", { name: "Select all" }),
+      ).not.toBeInTheDocument();
+    });
+
+    it("shows 'Deselect all' label when all rows are selected", async () => {
+      renderComponent(<DataTable {...initialProps} />);
+      const selectAll = screen.getByRole("checkbox", { name: "Select all" });
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx]: Why is this bound if it's only used once?

File: src/components/DataTable/DataTable.test.tsx [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+      await userEvent.click(selectAll);
+      expect(
+        screen.getByRole("checkbox", { name: "Deselect all" }),
+      ).toBeInTheDocument();
+    });
+
+    it("shows indeterminate state when some rows are selected", async () => {
+      renderComponent(<DataTable {...initialProps} />);
+      const rowCheckboxes = screen.getAllByRole("checkbox", {
+        name: /Select|Deselect/,
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx]: Check for specific variant, it should be deterministic. In this case, it'll be "Select" initially.

File: src/components/DataTable/DataTable.test.tsx [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+      renderComponent(
+        <DataTable {...initialProps} isRowDisabled={(row) => row.id === "1"} />,
+      );
+      const tableRows = screen.getAllByRole("row");
+      const aliceRow = tableRows.find((row) =>
+        row.textContent?.includes("Alice"),
+      );
+      const bobRow = tableRows.find((row) => row.textContent?.includes("Bob"));
+      expect(aliceRow).toHaveAttribute("aria-disabled", "true");
+      expect(bobRow).not.toHaveAttribute("aria-disabled", "true");
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx:161]: Use `toBeDisabled` here too

File: src/components/DataTable/DataTable.test.tsx [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        <DataTable {...initialProps} isRowDisabled={(row) => row.id === "1"} />,
+      );
+      const selectAll = screen.getByRole("checkbox", { name: "Select all" });
+      await userEvent.click(selectAll);
+      expect(screen.getByRole("checkbox", { name: "Select 2" })).toBeChecked();
+      expect(
+        screen.queryByRole("checkbox", { name: "Select 1" }),
+      ).not.toBeInTheDocument();
+    });
+  });
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx:175]: We need more tests for all the edge cases I have previously mentioned relating to the interaction between disabled/loading/selecting rows.

File: src/components/DataTable/DataTable.test.tsx [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+  describe("isRowLoading", () => {
+    it("excludes loading rows from select all", async () => {
+      renderComponent(
+        <DataTable {...initialProps} isRowLoading={(row) => row.id === "1"} />,
+      );
+      const selectAll = screen.getByRole("checkbox", { name: "Select all" });
+      await userEvent.click(selectAll);
+      expect(screen.getByRole("checkbox", { name: "Select 2" })).toBeChecked();
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx:184]: Actually need to check that the first row is not selected too.

File: src/components/DataTable/DataTable.test.tsx [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+describe("DataTable.columnBuilder", () => {
+  it("returns a column definition unchanged", () => {
+    const builder = DataTable.columnBuilder<{
+      id: string;
+      name: string;
+      age: number;
+    }>();
+    const col = builder({ header: "Name", map: (row) => row.name });
+    expect(col.header).toBe("Name");
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx]: we can check for identity here.

File: src/components/DataTable/useRows.test.tsx [hunk 19]
```diff
@@ -103,22 +104,60 @@ describe("useRows", () => {
...
+        createProps([{ key: "abc123", value: 123, otherValue: true }], {
+          selectable,
+          isRowLoading: () => loading,
+        }),
+      );
+
+      const row = result.current[0] as React.ReactElement<{
+        children: React.ReactNode[];
+      }>;
+      const [, childSelectable] = row.props.children;
```
Comment (andogq) [inline src/components/DataTable/useRows.test.tsx]:
```suggestion
      const [_, childSelectable] = row.props.children;
```
use a placeholder here to keep it clear.

**Review by andogq — CHANGES_REQUESTED** [review]
- Looks like there's lint issues that need to be addressed.
- Get rid of `is-dying` class, and use the `inert` attribute set on the actual table row. Then target the elements using that attribute in css.
- Try use the Vanilla table icon styling instead of manually creating types and overriding styles.

File: src/components/ModelTable/ModelTable.tsx (hunk 10 above)
Comment (Ninfa-Jeon) [reply src/components/ModelTable/ModelTable.tsx]: Pretty sure that was already there, good call out

File: src/components/DataTable/_data-table.scss (hunk 4 above)
Comment (Ninfa-Jeon) [reply src/components/DataTable/_data-table.scss:2]: Isn't that just for icon placeholders though?

File: src/components/DataTable/DataTable.tsx (hunk 5 above)
Comment (Ninfa-Jeon) [reply src/components/DataTable/DataTable.tsx:106]: This is also why I previously mentioned losing the `selectable` prop entirely. I will start work on the batch destruction coming week anyway.

File: src/components/InlineCheckbox/InlineCheckbox.tsx (hunk 9 above)
Comment (Ninfa-Jeon) [reply src/components/InlineCheckbox/InlineCheckbox.tsx:36]: That is because the `icon-column` class is making the icon header as well as the icon row cells appear well aligned with the other cells. PFA with/without my approach:
Screenshot 2026-08-20 at 8 22 05 PM Screenshot 2026-08-20 at 8 22 41 PM

Minor difference really, but it was pmo XD

File: src/components/ModelTable/ModelTable.tsx (hunk 11 above)
Comment (Ninfa-Jeon) [reply src/components/ModelTable/ModelTable.tsx]: This needed extra unrelated changes (request for raise)

File: src/components/ModelActions/ModelActions.test.tsx (hunk 12 above)
Comment (Ninfa-Jeon) [reply src/components/ModelActions/ModelActions.test.tsx:132]: If I remember correctly this assertion doesn't seem to work well for our codebase. Will try anyway

File: src/components/DataTable/DataTable.test.tsx (hunk 16 above)
Comment (Ninfa-Jeon) [reply src/components/DataTable/DataTable.test.tsx:175]: the tests I wrote are for extra coverage, we can improve it further in another PR :)

File: src/components/DataTable/_data-table.scss (hunk 4 above)
Comment (andogq) [reply src/components/DataTable/_data-table.scss:2]: From my testing, it seems to work for anything that is icon-like (particularly in sizing), including checkboxes and spinners. Wouldn't hurt to try, as it would be less for us to maintain.

File: src/components/DataTable/DataTable.test.tsx (hunk 15 above)
Comment (andogq) [reply src/components/DataTable/DataTable.test.tsx:161]: This selector still needs to be changed

File: src/components/DataTable/DataTable.test.tsx (hunk 16 above)
Comment (andogq) [reply src/components/DataTable/DataTable.test.tsx:175]: That's fine, but it's good practice to have tests covering the features that are added in a PR. There shouldn't be many needed, maybe 4 or 5 to cover the edge cases we've discussed.

File: src/components/DataTable/DataTable.test.tsx [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+    const builder = DataTable.columnBuilder<{
+      id: string;
+      name: string;
+      age: number;
+    }>();
+    const mapFn = (row: { id: string; name: string; age: number }): string =>
+      row.name;
+    const col = builder({ header: "Name", map: mapFn });
+    expect(col.header).toBe("Name");
+    expect(col.map).toBe(mapFn);
```
Comment (andogq) [inline src/components/DataTable/DataTable.test.tsx]:
```suggestion
    const colRaw = { header: "Name", map: mapFn };
    const col = builder(colRaw);
    expect(col).toBe(colRaw);
```

Since the builder function just returns whatever it was passed (it's the identity function), we can perform a direct comparison between the two objects.

**Review by andogq — APPROVED** [review]
Nearly there. A few more things that you need to do, so please make sure they're done before merging.
- Test the usage of the vanilla icon class, it'd be very convenient if it works. If it doesn't work, then that's ok.
- Use the `isDisabled` attribute in tests
- Add missing tests for the edge cases we've discussed. Since they're unexpected and tricky to reason about, tests are important to ensure that whatever behaviour we end with is consistent and repeatable.

File: src/components/ModelActions/ModelActions.test.tsx (hunk 12 above)
Comment (Ninfa-Jeon) [reply src/components/ModelActions/ModelActions.test.tsx:132]: As stated above, I can confirm that this assertion doesn't work on our codebase as it doesn't take into account the `aria-disabled` attribute when checking for disabled state.

File: src/components/DataTable/DataTable.test.tsx (hunk 15 above)
Comment (Ninfa-Jeon) [reply src/components/DataTable/DataTable.test.tsx:161]: Doesn't work with `toBeDisabled` selector.

File: src/components/DataTable/_data-table.scss (hunk 4 above)
Comment (Ninfa-Jeon) [reply src/components/DataTable/_data-table.scss:2]: The column doesn't look right to me when the class is replaced with `.p-table__cell--icon-placeholder`. Maybe I am not understanding what you're asking for. I will follow up with another PR for this once you explain it to me.
